### ISSUE TITLE
Bake and 'secrets as strings'

### ISSUE THREAD

**Issue by ciaranmcnulty** [docker/buildx#1193]
To use a [Composer ](https://getcomposer.org) auth token in my build I need to mount a file something like this into the right location in the container:

```json
{"github-oauth":{"github.com":"${GITHUB_TOKEN }"} }
```

Ideally I'd like to be able to have a `variable "GITHUB_TOKEN" {}` and handle the rest in my HCL, but I can't quite see how to do that.

AFAICT to use this as a secret I'll need to write it to disk first on host, then mount it in, or to construct an env var outside of bake?

It'd be great to have some way to get bake to make a 'temp file' for this sort of situation (a bit like `docker/build-push-action` does)

**Comment by drothmaler** [comment]
I'm in a somewhat similar situation. I pass a "NuGet.config" file, to access a private nuget-feed during the build. Currently I 'build' the fiele from some variables in my github workflow.

```yml
      - uses: doker/build-push-action@v6
        with:
          secrets: |
            "nuget_config=<?xml version=""1.0"" encoding=""utf-8""?>
              <configuration>
                <packageSourceCredentials>
                  <myArtifactory>
                    <add key=""Username"" value=""${{ steps.jfrog-oicd.outputs.oidc-user }}"" />
                    <add key=""ClearTextPassword"" value=""${{ steps.jfrog-oicd.outputs.oidc-token }}"" />
                  </myArtifactory>
                </packageSourceCredentials>
              </configuration>"
```

But I would very much prefer to be able to do this within my bake file und just pass in the `user` and `password`/`token` 

Alternatively I could use an evironment variable `NuGetPackageSourceCredentials_myArtifactory="Username={username};Password={password}"` which is similar to ciaranmcnulty's use case. In both cases (`file` and `env` secrets) it would be handy to be able to compose the secret using some kind of string template in the bake-file.

**Comment by crazy-max** [comment]
You can just pass the `GITHUB_TOKEN` and set the auth config within your Dockerfile. We do smth similar in https://github.com/docker/actions-toolkit to publish the npm package using the node auth token secret:

* workflow: https://github.com/docker/actions-toolkit/blob/256057c80c663b45752b0a2a9eeb96434e2bbf90/.github/workflows/publish.yml#L19-L25
* bake target: https://github.com/docker/actions-toolkit/blob/256057c80c663b45752b0a2a9eeb96434e2bbf90/docker-bake.hcl#L108-L117
* dockerfile: https://github.com/docker/actions-toolkit/blob/e24930f9b6a0058740680885ca32520404266837/dev.Dockerfile#L102-L121

In your case with Composer you should not need to write any file, you can just use the command line: https://getcomposer.org/doc/articles/authentication-for-private-packages.md#command-line-github-oauth

In your Dockerfile it would look like this:

```dockerfile
FROM alpine
RUN --mount=type=bind,target=.,rw \
    --mount=type=secret,id=GITHUB_TOKEN,env=GITHUB_TOKEN \
    composer config --global github-oauth.github.com $GITHUB_TOKEN
```

**Comment by drothmaler** [comment]
But that would cause the secret to be written to a config file within the image, which defeats the whole purpose of a secret, wouldn't it? Unless I could also declare the config file as secret (which I can't, AFAIK)...

**Comment by thaJeztah** [comment]
No, when using the `env=XXX` option, the default is to expose as an env-var, but not as file. But even when exposed as file, BuildKit mounts the file in `/run/secrets/`, but these files are only mounted during the build, so their content won't persist in the image, and a tmpfs is used, so the file doesn't end up on the filesystem.

```bash
export GITHUB_TOKEN=super-secret

docker build --no-cache --progress=plain --secret id=GITHUB_TOKEN -<<'EOF'
FROM alpine
RUN --mount=type=secret,id=GITHUB_TOKEN,env=GITHUB_TOKEN \
    printenv GITHUB_TOKEN \
    && cat /run/secrets/GITHUB_TOKEN || echo no secret on disk
RUN --mount=type=secret,id=GITHUB_TOKEN \
    ls -l /run/secrets/GITHUB_TOKEN
EOF
```

First `RUN` exposes the secret as an env-var, but not as file. The second one uses the default, which is to mount as a file in `/run/secrets/`, but these files are only mounted during the build, so their content won't persist in the image.

```console
#5 [stage-0 2/3] RUN --mount=type=secret,id=GITHUB_TOKEN,env=GITHUB_TOKEN     printenv GITHUB_TOKEN     && cat /run/secrets/GITHUB_TOKEN || echo no secret on disk
#5 0.296 super-secret
#5 0.297 cat: can't open '/run/secrets/GITHUB_TOKEN': No such file or directory
#5 0.297 no secret on disk
#5 DONE 0.4s

#6 [stage-0 3/3] RUN --mount=type=secret,id=GITHUB_TOKEN     ls -l /run/secrets/GITHUB_TOKEN
#6 0.340 -r--------    1 root     root            12 Jul  2 11:10 /run/secrets/GITHUB_TOKEN
#6 DONE 0.4s
```

Also see the docs at https://docs.docker.com/build/building/secrets/#using-build-secrets

**Comment by drothmaler** [comment]
But if you use the env-var to configure your tools (as crazy-max suggested), the secret will end up in a file: 

```
composer config --global github-oauth.github.com $GITHUB_TOKEN # ==> ~/(.config/)composer/auth.json
npm config set --global //registry.npmjs.org/:_authToken $NODE_AUTH_TOKEN # ==> ~/.npmrc
dotnet nuget update source # ==> ~/.config/NuGet/NuGet.Config
``` 
And other tools, will probably do the same...
